rohmu/transfer_pool.py

```diff
@@ -0,0 +1,220 @@
+from .errors import InvalidTransferError
+from .object_storage.base import BaseTransfer, StorageModel
+from contextlib import contextmanager
+from typing import Any, Callable, Generator, Optional
+
+import heapq
+import json
+import logging
+import threading
+import time
+
+LOG = logging.getLogger(__name__)
+
+TRANSFER_MAX_AGE = 60 * 60 * 3
+TRANSFER_CACHE_MAX_AGE = TRANSFER_MAX_AGE * 2
+
+
+class TransferCacheItem:
+    create_time: float
+    cache_key: str
+    transfer: BaseTransfer[StorageModel]
+    max_age: float
+
+    def __init__(self, pool_key: str, transfer: BaseTransfer[StorageModel], max_age: float = TRANSFER_MAX_AGE) -> None:
+        self.create_time = time.monotonic()
+        self.cache_key = pool_key
+        self.transfer = transfer
+        self.max_age = max_age
+
+    def age(self) -> float:
+        return time.monotonic() - self.create_time
+
+    def is_expired(self) -> bool:
+        return self.age() > self.max_age
+
+    def __gt__(self, other: TransferCacheItem) -> bool:
+        """Ordering by creation time, needed for storing these in a heap."""
+        return self.create_time > other.create_time
+
+
+class _TransferCache:
+    transfers_heap: list[TransferCacheItem]
+    last_used: float
+    max_age: float
+
+    def __init__(self, max_age: float = TRANSFER_CACHE_MAX_AGE) -> None:
+        self.transfers_heap = []
+        self.last_used = time.monotonic()
+        self.max_age = max_age
+
+    def prune_expired(self) -> None:
+        while self.transfers_heap and self.transfers_heap[0].is_expired():
+            heapq.heappop(self.transfers_heap)
+
+    def get(self) -> Optional[TransferCacheItem]:
+        self.last_used = time.monotonic()
+        self.prune_expired()
+        if self.transfers_heap:
+            try:
+                return heapq.heappop(self.transfers_heap)
+            except IndexError:
+                pass
+        return None
+
+    def put(self, transfer: TransferCacheItem) -> None:
+        self.last_used = time.monotonic()
+        if not transfer.is_expired():
+            heapq.heappush(self.transfers_heap, transfer)
+
+    def age(self) -> float:
+        return time.monotonic() - self.last_used
+
+    def is_expired(self) -> bool:
+        return self.age() > self.max_age
+
+
+class _TransferCacheForThreadSafeTransfer(_TransferCache):
+    """This _TransferCache subclass is a simple implementation that works when the transfer is thread-safe.
+
+    In that case we can have a single instance and always return that one.
+
+    We still apply the expiration logic though.
+    """
+
+    def get(self) -> Optional[TransferCacheItem]:
+        self.last_used = time.monotonic()
+        self.prune_expired()
+        if self.transfers_heap:
+            try:
+                # there will always be only one transfer instance at most in this cache type
+                return self.transfers_heap[0]
+            except IndexError:
+                pass
+        return None
```

Comment:
```suggestion
    """

    def get(self) -> Optional[TransferCacheItem]:
        self.last_used = time.monotonic()
        if self.transfers_heap:
            try:
                # there will always be only one transfer instance at most in this cache type
                return self.transfers_heap[0]
            except IndexError:
                pass
        return None
```

Expiration of transfers in single transfer case isn't really meaningful. For multi transfer cache, it matters, but here we'd be recreating identical object and that is just waste of resources (presumably per-request stuff isn't long lived anyway, and even if it was, deleting it just before recreating it probably won't help much).

Reply:
Removed expiration in that pool and made sure we always have a single instance.